Day 1, 2, 3, & 4 MVP

Catherinesjkim opened the pull request:
Because:
- Implemented a basic hash table without collision resolution
- Implemented a HashTable class and HashTableEntry class
- Implemented a good hashing function with djb2
- Implemented the hash_index() that returns an index value for a key
- Implemented the put(), get(), and delete() methods.

Catherinesjkim committed:
Add Updated Files
Because:
- Added some code
- Not sure if the tests are passing
- Will check tomorrow AM.

Catherinesjkim committed:
Add Day 1
Because:
- Implemented a basic hash table without collision resolution
- Implemented a HashTable class and HashTableEntry class
- Implemented a good hashing function with djb2
- Implemented the hash_index() that returns an index value for a key
- Implemented the put(), get(), and delete() methods.

Catherinesjkim committed:
Add Part 1 MVP
Because:
- Refactored the code to output the values instead of just the keys.

Catherinesjkim committed:
Add Updated Hashes File
Because: Updated Artem video lecture notes
- test case test_hashtable_no_collisions.py passed for hashtable.py.

LorenzoEvans commented on hashes.py:
```diff
@@ -0,0 +1,128 @@
+    # take every character in the string, and convert character to number
+    # Convert each character into UTF-8 numbers (it allows for more characters than ASCII & Unicode)
+    string_utf = self.encode()
+    
```
Awesome move, converting to utf-8, it's good to be aware of that string format, as well as making sure you can handle a wider range of strings and characters- you never know what someone will want to use as a key.

Catherinesjkim replied:
Yeap, specially these days, we do a lot of transactions/businesses with China and other countries : )

LorenzoEvans commented on hashes.py:
```diff
@@ -0,0 +1,128 @@
+# Non-reversible
+# Avoid collisions
+
+hash_table = [None] * 8 # 8 slots, all initialized to None
```
Just think it's cool that you chose the function application method as opposed to the class based method for implementing the HT. When you find yourself in a technical interview, always ask if you can do your data structures like this, because sometimes it's just easier to think about applying functions to data, versus using a method on an object to change the object (smells recursive, no?).

LorenzoEvans commented on hashtable/hashtable.py:
```diff
@@ -7,22 +8,25 @@ def __init__(self, key, value):
-
+"""
+# HashTable class using chaining
+class HashTable:
```
Class implementation- W00t! Some employers will let you use one way, and then ask how you'd do it _the other_ way, so it's great that you're aware of both!

LorenzoEvans commented on leetCode.py:
```diff
@@ -0,0 +1,25 @@
+O(n): n == length of the S
+O(length of j)
+"""
+class Solution:
```
Love that you're adding external resources into your project and learning! Yes!

Type hints are a very powerful tool in Python, that allows it to extend a little into true functional programming (typically there's a type system that will allow your compiler (though Python is _interpreted_), to predict what your code should and shouldn't be doing _without you explicitly stating it_), and it's just great to have as a reference.

Catherinesjkim committed:
Add Updated Files
Because:
- Added shrink method to the get load factor and resize functions.

Catherinesjkim committed:
Add Day 2 MVP Met
Because:
- Implemented LL chaining for collision resolution

This commmit:
- Modified put, get, and delete methods to handle collisions
- Implemented load factor measurements and automatic hastable size doubling
- test_hashtable_resize file passed.

Catherinesjkim committed:
Add no_dups file
Because:
- Output the string in the same order but with subsequent dupe words removed
- There's no extra spaces at the end of my returned string but there's a None
- Will check runtime complexity soon.

Catherinesjkim committed:
Add word_count file
Because:
- Implemented a word count fn that takes a single string as an argument
- It returns a dictionary of words and their counts
- Case is ignored. Output keys are in lowercase
- Key order in the dictionary didn't matter
- Spllit the strings into words on any whitespace
- Will work on ignoring each of the special characters soon.

Catherinesjkim committed:
Add Updated no_dups and hashtable
Because:
- Added return s for the output to only show the string in the same order without the subsequent dupe words removed.

Catherinesjkim committed:
Add Updated Word Count File
Because:
- Refatored the code so that the word count fn ignores each of the special characters.

Catherinesjkim committed:
Add Histo File
Because:
- The fn takes a single filename string as an argument robin.txt
- Opens the file, works through it to produce an output

This commit:
- Prints a histogram showing the word count for each word in integers
- Working on ordering by the number of words, then by the word (alphabetically).

Catherinesjkim committed:
Add Expendive Sequence File
Because:
- Computed the Expensive Sequence
- First implementation without the hashtable.

Catherinesjkim committed:
Add Updated Expensive Sequence File
Because:
- Refactored the code using a hastable to make sure my solution completes before the universe ends.

Catherinesjkim committed:
Add Markov File
Because:
- There is an output after inputting a word
- Not sure if it is working correctly. Will check tomorrow.
